### PR TITLE
WIP: Begin crate smashing

### PR DESCRIPTION
**Pull request by tcharding**
Not a merge candidate until release 0.29 is done. Throwing it up for early feedback please.

This is my most recent attempt to understand the 'crate smashing' [thread](https://github.com/rust-bitcoin/rust-bitcoin/issues/550) and split out a few crates. Specifically this PR currently creates the following crates, as a workspace:
- bitcoin: main crate
- bitcoin-str: hex, and re-exports (optional dependencies) of bech32, and base64
- endian: the endian module turned into a crate
- base58: the base58 module turned into a crate

It all builds locally but I doubt I have CI sorted correctly yet.

### CODE REVIEW

File: bitcoin-str/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+/// Re-export of the [`base64`] crate.
+/// [`base64`]: <https://docs.rs/base64/latest/base64/>
+#[cfg(feature = "base64")]
+pub use base64;
+
+// Only warn for std builds, if devs are doing no-std builds they probably know what they are doing.
+#[cfg(all(feature = "std", feature = "bech32", not(feature = "bech32-std")))]
+compile_error!("If you enable \"std\" and want \"bech32\" you should enable \"bech32-std\"");
+#[cfg(all(feature = "std", feature = "base64", not(feature = "base64-std")))]
+compile_error!("If you enable \"std\" and want \"base64\" you should enable \"base64-std\"");
```
Comment (Kixunil) [inline bitcoin-str/src/lib.rs:51]: If I'm not mistaken these make cargo features non-additive

File: endian/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+[package]
```
Comment (Kixunil) [inline endian/Cargo.toml:1]: This looks like `bitcoin-internal-macros` would be a better place for it.

File: bitcoin/src/util/address.rs [hunk 3]
```diff
@@ -30,13 +30,13 @@ use crate::error::ParseIntError;
 use core::str::FromStr;
 
 use secp256k1::{Secp256k1, Verification, XOnlyPublicKey};
-use bech32;
```
Comment (Kixunil) [inline bitcoin/src/util/address.rs:33]: Can we get rid of `util` nesting? Maybe even before smashing?

**Review by Kixunil — COMMENTED** [review]
Took a very light look, some ideas below.

Thoughts on internal-macros strategy: naive management would cause duplicate dependencies but if we ever release a breaking version we could update all dependents to use the new one. So if people update all relevant dependencies they won't get duplicates. I think this is auto-checkable.

File: bitcoin-str/src/lib.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin-str/src/lib.rs:51]: oooo yes, I didn't think of that. Thanks

**Comment by tcharding** [comment]
Stale, closing.

**Pull request closed without merging**